### PR TITLE
chore: detect the channel a PR wants to merge into

### PR DESCRIPTION
**Pull request by weihanglo**
### What does this PR try to resolve?

This detects which channel a pull request wants to merge into.

It is hard because bors runs in a different repo.
Bors knows nothing about branches or tag in this repo.
It only see one base commit and a merge commit AFAIK.

Basically the assumption and idea are

1. bors create a merge commit, so `HEAD~` should find the base branch.
2. Cargo maintains `rust-1.y.0` branch for each Rust minor version releases.
   Therefore, we can use the symbolic name of `origin/rust-1.x.0`
   to determine if it aims for stable, beta, or nightly channel.
3. After we know which channel it is targeted at,
   we can skip jobs that are likely to be broken.

### How should we test and review this PR?

I setup a rust-1.70.0 branch on ec8a8a0cabb0e0cadef58902470f6c7ee7868bdc on my repo. 
The script does find beta branch correctly.
https://github.com/weihanglo/cargo/actions/runs/5081469788/jobs/9129891293#step:3:36

### Additional information

Sorry for bringing more shell scripts 🤯.

If this looks good, I'll file a PR for master branch.

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -57,82 +73,106 @@ jobs:
...
+          if: always()
         - name: Windows x86_64 gnu nightly # runs out of space while trying to link the test suite
           os: windows-latest
           rust: nightly-gnu
           other: i686-pc-windows-gnu
+          if: ${{ needs.channel.outputs.CHANNEL == 'nightly' }}
     name: Tests ${{ matrix.name }}
     steps:
     - uses: actions/checkout@v3
+      if: matrix.if
```
Comment (weihanglo) [inline .github/workflows/main.yml:110]: Unfortunately `job.if` cannot access fields in `matrix`. We cannot simply write `if: matrix.if` for `job` and instead we need to do it in each `step`. That's terrible. Totally understand if someone objects to this PR due to that.

See https://github.com/actions/runner/issues/1985

File: ci/which-channel.sh [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+# Default to nightly channel
+CHANNEL="nightly"
+
+# Get symbolic names for the base_sha.
+# We only care about `rust-1.y.z` branches
+ref=$(git name-rev --name-only --refs="origin/rust-1.*" $base_sha)
+
+# Remove the patch version `.0` and keep only `1.y`,
+# so that rustup can install the correct patched toolchain
+if [[ "$ref" =~ ^origin/rust-(.*).0$ ]]
```
Comment (ehuss) [inline ci/which-channel.sh:25]: I'm a bit uncertain about how this works, perhaps you can clarify.

Locally running `git name-rev`, my output looks like `origin/rust-1.69.0~1` where there is a `~` character. But the regex here doesn't allot for that. I'm a bit confused by that.

Also, does this work if I open a PR to a beta branch, but my commit is a few commits behind the tip of the branch?

File: .github/workflows/main.yml (hunk 1 above)
Comment (ehuss) [reply .github/workflows/main.yml:110]: Ugh, that is really unfortunate.  Is it possible to use something like https://stackoverflow.com/a/65434401 to have a job that runs first to generate the matrix? I'm not sure if that's any better, since it is also a pretty complex solution.

File: ci/which-channel.sh (hunk 2 above)
Comment (weihanglo) [reply ci/which-channel.sh:25]: The assumption here is that a `base_sha` is always HEAD of`rust-1.*.0` branch. It either gets the commit from `github.event.pull_request.base.sha`, or from bors `merge-commit~1`. If this assumption is correct, I think `git name-rev` will always get a ref without `~` or `^`.

File: ci/which-channel.sh [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+    if [[ "$beta" == "rustc ${version}"* ]]
+    then
+        CHANNEL="beta"
+    fi
+
+    if [[ "$stable" = "rustc ${version}"* ]]
+    then
+        CHANNEL="stable"
+    fi
```
[the remark this answers is not in the document]
Comment (weihanglo) [reply ci/which-channel.sh:41]: Either way looks more robust to me. I'll try the latter since it reduces  one toolchain download.

File: ci/which-channel.sh (hunk 2 above)
Comment (weihanglo) [reply ci/which-channel.sh]: Fixed with 6a973cf754fe4f7596656f203902c81427111453.

File: ci/which-channel.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (weihanglo) [reply ci/which-channel.sh]: Fixed with 6a973cf754fe4f7596656f203902c81427111453.

File: .github/workflows/main.yml (hunk 1 above)
Comment (weihanglo) [reply .github/workflows/main.yml]: Fixed with 4ff82fdc391a2b06ae3c9bc4630e0314428a5a18. Doesn't look too bad though.

File: ci/which-channel.sh [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+base_sha=$(git rev-parse "${BASE_SHA:-HEAD~1}")
+
+# Get symbolic names for the base_sha.
+# Assumption: Cargo branches are always in the format of `rust-1.*.0`,
+#             otherwise `git name-rev` will return "undefined".
+ref=$(git name-rev --name-only --refs='origin/rust-1.*.0' $base_sha)
+
+# Get the latest `rust-1.*.0` branch from remote origin.
+# Assumption: The latest branch is always beta branch.
+beta=$(git branch --remotes --list 'origin/rust-1.*.0' | sort | tail -n1 | tr -d "[:space:]")
```
Comment (ehuss) [inline ci/which-channel.sh:23]: This looks like it might have some issues with three-digit version like 1.100. Perhaps it could be something like `sed -E 's/ *upstream\/rust-1.([0-9]+).0/\1/g' | sort -n`?

File: ci/which-channel.sh [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    # Should be nightly but for convenience in CI let's call it master.
+    channel="master"
+else
+    if [[ "$ref" = "$beta" ]]
+    then
+        channel="beta"
+    else
+        channel="stable"
+    fi
+fi
```
Comment (ehuss) [inline ci/which-channel.sh:39]: I think this is maybe a little simpler (though double-check, I believe this is correct).

```suggestion
elif [[ "$ref" = "$beta" ]]
then
    channel="beta"
else
    channel="stable"
fi
```

File: .github/workflows/main.yml [hunk 6]
```diff
@@ -41,8 +55,33 @@ jobs:
...
+    - name: Generate strategy matrix
+      id: matrix
+      run: |
+        CHANNEL=${{ needs.channel.outputs.CHANNEL }}
+        MATRIX=$(
+          jq --arg C "$CHANNEL" 'map (. | 
+            if ($C == "beta") then select(.rust | startswith("nightly") | not)
+            elif ($C == "stable") then select(.rust | startswith("stable"))
+            else . end)' ci/matrix.json
+        )
```
Comment (ehuss) [inline .github/workflows/main.yml:77]: Can this have a comment explaining roughly what it is doing? It's not too hard to follow, but just a high-level description could help.

File: ci/which-channel.sh (hunk 4 above)
Comment (weihanglo) [reply ci/which-channel.sh]: I went using `sort --version-sort`. Tested on ubuntu-latest (22.04). It can correctly sort the following versions from

```
origin/rust-1.100.0
origin/rust-1.50.0
origin/rust-1.222.0
origin/rust-1.48.0
origin/rust-1.70.0
origin/rust-1.60.0
origin/rust-1.69.0
origin/rust-1.9.0
```
to

```
origin/rust-1.9.0
origin/rust-1.48.0
origin/rust-1.50.0
origin/rust-1.60.0
origin/rust-1.69.0
origin/rust-1.70.0
origin/rust-1.100.0
origin/rust-1.222.0
```

File: .github/workflows/main.yml (hunk 6 above)
Comment (weihanglo) [reply .github/workflows/main.yml:84]: Added. Let me know if it is clear enough.

**Comment by weihanglo** [comment]
Thanks for the review! I rebased the pull request. Change diff can be seen from [here](https://github.com/rust-lang/cargo/compare/43ae7846490a98df1020162b4b5cbd0deac09c80..11704cbf54beb2348529f91bd9199d9616246d4e).

**Comment by ehuss** [comment]
Thanks!  I'm not sure if it matters much that we're merging this to rust-1.70.0, but I think it should be fine.  Do you also want to backport this to rust-1.71.0?

I kinda wish GitHub Actions would have a little more diagnostics or descriptions on the runs page to explain why a job was skipped, but hopefully it will be clear to people in the future.

@bors r+

**Comment by weihanglo** [comment]
Sure for 1.71.0.

Now it sounds like we need two more PRs — one for 1.71 beta and the other for master 😂

**Comment by ehuss** [comment]
Oops, looks like skipped jobs cause a failure in the `if` condition. 
See: https://github.com/orgs/community/discussions/45058 and https://github.com/actions/runner/issues/491.
Are there any workarounds listed there that might work?

**Comment by weihanglo** [comment]
Ok. `build-std` job was skipped so bors failed. Maybe we need a weak depedency syntax for `needs`, like 

```
needs: ["build-std?", "test"]
```

Just kidding.

**Comment by weihanglo** [comment]
Something like this? I can send PRs to 1.71.0 and master and then close this if this patch seems good.

```patch
diff --git a/.github/workflows/main.yml b/.github/workflows/main.yml
index 2f75f206b..5f955af89 100644
--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -227,11 +228,16 @@ jobs:
           "https://raw.githubusercontent.com/rust-lang/rust/$BRANCH/src/tools/linkchecker/linkcheck.sh"
         sh linkcheck.sh --all cargo
 
+  nightly-only-jobs:
+    needs: [build_std]
+    if: "(jobs.channel.outputs.CHANNEL != 'master' && contains(needs.*.result, 'skipped')) || success()"
+    runs-on: ubuntu-latest
+
   success:
     permissions:
       contents: none
     name: bors build finished
-    needs: [docs, rustfmt, test, resolver, build_std, test_gitoxide]
+    needs: [docs, rustfmt, test, resolver, nightly-only-jobs, test_gitoxide]
     runs-on: ubuntu-latest
     if: "success() && github.event_name == 'push' && github.ref == 'refs/heads/auto-cargo'"
     steps:
@@ -240,7 +246,7 @@ jobs:
     permissions:
       contents: none
     name: bors build finished
-    needs: [docs, rustfmt, test, resolver, build_std]
+    needs: [docs, rustfmt, test, resolver, nightly-only-jobs]
     runs-on: ubuntu-latest
     if: "!success() && github.event_name == 'push' && github.ref == 'refs/heads/auto-cargo'"
     steps:
```

**Comment by weihanglo** [comment]
Okay I've updated this to target on the current master. We can have a new round of review :)
